### PR TITLE
Feature/cross entropy tests

### PR DESCRIPTION
**Pull request by baykaner**
- tests/bugfixes/refactoring for cross entropy/softmax_cross_entropy/scaled_cross_entropy
- bugfix for Argmax which assumed tensors all 1-dim
- implementing first ml integration tests checking training/convergence in very basic scenarios
- minor tweaks/improvement to SGNS example

### CODE REVIEW

File: libs/core/include/core/fixed_point/fixed_point.hpp [hunk 1]
```diff
@@ -350,22 +350,32 @@ class FixedPoint
...
+  {
+    return (static_cast<unsigned>(data_) / one);
+  }
+
+  explicit operator unsigned long() const
+  {
+    return (static_cast<unsigned long>(data_) / one);
+  }
+
+  explicit operator unsigned long long() const
```
Comment (pierre-wilmot-fetch-ai) [inline libs/core/include/core/fixed_point/fixed_point.hpp:378]: Does that makes sense ? 
What's the expected result when the FixedPoint is representing a negative value ?
In what situation do we need that ?

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp [hunk 2]
```diff
@@ -544,26 +544,60 @@ void PeakToPeak(ArrayType arr)
  * @return
  */
 template <typename ArrayType>
-void ArgMax(ArrayType const &array, typename ArrayType::SizeType &ret)
+void ArgMax(ArrayType const &array, ArrayType &ret, typename ArrayType::SizeType axis = 0)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:547]: Why is ArgMax not returning a SizeType anymore ?

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp [hunk 3]
```diff
@@ -544,26 +544,60 @@ void PeakToPeak(ArrayType arr)
...
+  {
+    // 2D argmax result has size
+    if (axis == 0)
+    {
+      ret = ArrayType{{array.shape().at(axis), typename ArrayType::SizeType(1)}};
+    }
+    else
+    {
+      ret = ArrayType{{typename ArrayType::SizeType(1), array.shape().at(axis)}};
+    }
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:606]: Processing for axis == 1 is most likely broken

File: libs/math/include/math/free_functions/standard_functions/log.hpp [hunk 4]
```diff
@@ -70,25 +77,27 @@ meta::IfIsNonBlasArray<ArrayType, void> Log(ArrayType &x)
...
 template <typename T>
 meta::IfIsMathArray<T, void> Log(T const &array, T &ret)
 {
   ret = array;
-  for (typename T::Type &e : ret)
+  typename T::SizeType ret_count{0};
+  for (typename T::Type &e : array)
   {
-    Log(e);
+    Log(e, ret.At(ret_count));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/standard_functions/log.hpp:102]: Since ret = array (L95), you're effectivily overwriting your input.

File: libs/math/tests/math/ml_loss_functions/cross_entropy.cpp [hunk 5]
```diff
@@ -105,6 +105,5 @@ TYPED_TEST(CrossEntropyTest, value_test)
   score = fetch::math::CrossEntropyLoss(test_array, gt_array);
 
   // test correct values
-  ASSERT_NEAR(score, typename TypeParam::Type(1.78777539730072021484375),
-              typename TypeParam::Type(1.0e-5f));
+  ASSERT_NEAR(double(score), double(0.893887639), double(1.0e-5f));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/tests/math/ml_loss_functions/cross_entropy.cpp:108]: Where are these values coming from ? 
That's the second time I see ground thruth values for math function unit test being modified even though the input is the same.

File: libs/ml/examples/file_loader.hpp [hunk 6]
```diff
@@ -42,7 +48,7 @@ std::vector<std::string> GetAllTextFiles(std::string const &dir_name)
       p1 = strtok(NULL, ".");
       if (p1 != NULL)
       {
-        txt_cmp = strcmp(p1, "txt");
+        txt_cmp = std::strcmp(p1, "txt");
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/file_loader.hpp]: I still think this C-code looks very out of place in out C++ codebase

File: libs/ml/examples/model_saver.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace fetch {
+namespace ml {
+namespace examples {
+
+/**
+ * Saves the state dict of a graph to a file location specified by user
+ * @param g the graph to save
+ * @param save_location a string specifying save location
+ * @return bool specifiy succes
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/model_saver.hpp]: function has no return value

File: libs/ml/examples/model_saver.hpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ */
+template <typename GraphType>
+void SaveModel(GraphType g, std::string const &save_location)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/model_saver.hpp]: GraphType g => GraphType const &g.
No need to make a copy

File: libs/ml/examples/word2vec_SGNS/main.cpp [hunk 9]
```diff
@@ -42,33 +48,34 @@ using SizeType     = typename ArrayType::SizeType;
...
-  SizeType    k               = 10;      // how many nearest neighbours to compare against
-  std::string test_word       = "cold";  // test word to consider
+  SizeType    output_size     = 1;
+  SizeType    batch_size      = 128;            // training data batch size
+  SizeType    embedding_size  = 16;             // dimension of embedding vec
+  SizeType    training_epochs = 100000;         // total number of training epochs
+  double      learning_rate   = 0.1;            // alpha - the learning rate
+  SizeType    k               = 10;             // how many nearest neighbours to compare against
+  std::string test_word       = "cold";         // test word to consider
+  std::string save_loc        = "./model.fba";  // small value for avoiding numerical instability
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/word2vec_SGNS/main.cpp]: Out of sync comment

File: libs/ml/include/ml/ops/activations/sigmoid.hpp [hunk 10]
```diff
@@ -45,6 +46,14 @@ class Sigmoid : public fetch::ml::ElementWiseOps<T>
     }
 
     fetch::math::Sigmoid(inputs.front().get(), *this->output_);
+
+    // ensures numerical stability
+    for (auto &val : *this->output_)
+    {
+      fetch::math::Max(val, epsilon_, val);
+      fetch::math::Min(val, fetch::math::Subtract(DataType(1), epsilon_), val);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/ops/activations/sigmoid.hpp:54]: Let maybe add a fetch::math::Clamp(T &val, Tconst &min, T const &max)

File: libs/ml/tests/ml/ops/activations/sigmoid.cpp [hunk 11]
```diff
@@ -58,8 +58,7 @@ TYPED_TEST(SigmoidTest, backward_test)
   TypeParam           gt(8);
   std::vector<double> dataInput({1, -2, 3, -4, 5, -6, 7, -8});
   std::vector<double> errorInput({0, 0, 0, 0, 1, 0, 0, 0});
-  std::vector<double> gtInput({-0.196612, -0.104994, -0.0451767, -0.0176627, 0.993352, -0.00246651,
-                               -0.000910167, -0.000335238});
+  std::vector<double> gtInput({0, 0, 0, 0, 0.00664803, 0, 0, 0});
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/ops/activations/sigmoid.cpp:61]: Where are these value coming from ?
Where were the previous values coming from ?

File: libs/ml/tests/ml/ops/loss_functions/cross_entropy.cpp [hunk 12]
```diff
@@ -96,40 +96,136 @@ TYPED_TEST(CrossEntropyTest, one_dimensional_forward_test)
   }
 
   fetch::ml::ops::CrossEntropy<TypeParam> op;
-  ASSERT_FLOAT_EQ(float(op.Forward({data1, data2})), 1.78777539730072021484375);
+  ASSERT_FLOAT_EQ(float(op.Forward({data1, data2})), float(0.893887639));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/ops/loss_functions/cross_entropy.cpp:99]: Where are these value coming from ?
Where were the previous values coming from ?

File: scripts/python_ml_lib/cross_entropy_test.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
```
Comment (pierre-wilmot-fetch-ai) [inline scripts/python_ml_lib/cross_entropy_test.py:1]: Leftover test code ?

File: libs/ml/tests/CMakeLists.txt [hunk 14]
```diff
@@ -13,5 +13,5 @@ add_fetch_test(ml_ops_gtest fetch-ml ml/ops)
 add_fetch_test(ml_serializers_gtest fetch-ml ml/serializers)
 add_fetch_test(ml_dataloader_gtest fetch-ml ml/dataloader)
 add_fetch_test(ml_layers_gtest fetch-ml ml/layers)
-
+add_fetch_test(ml_training_gtest fetch-ml ml/training)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/CMakeLists.txt]: Let's mark these est as integration
Test #36: ml_training_gtest .................   Passed   16.59 sec

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

File: libs/core/include/core/fixed_point/fixed_point.hpp (hunk 1 above)
Comment (baykaner) [reply libs/core/include/core/fixed_point/fixed_point.hpp:378]: as discussed argmax or indexing in general requires DataType -> SizeType conversions

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp (hunk 2 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:547]: so that ArgMax can handle a 2D matrix of for example (n_data x n_features) and return a vector of arg maxes. If you want argmax over the whole 2d matrix I guess it should be flatten first or call argmax on the resultant vector.

Do you think a different interface should be preferred for handling both argmax over entire 2d tensor AND batch argmax?

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp (hunk 3 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:606]: yep nice catch

File: libs/math/include/math/free_functions/standard_functions/log.hpp (hunk 4 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/standard_functions/log.hpp:102]: yep, and the two log functions defined above have the wrong interface too. fixed

File: libs/math/tests/math/ml_loss_functions/cross_entropy.cpp (hunk 5 above)
Comment (baykaner) [reply libs/math/tests/math/ml_loss_functions/cross_entropy.cpp:108]: python script uploaded showing how to calculate values in numpy for future reference if necessary

File: libs/ml/include/ml/ops/activations/sigmoid.hpp (hunk 10 above)
Comment (baykaner) [reply libs/ml/include/ml/ops/activations/sigmoid.hpp:54]: yes, this is going to be an extremely resuable piece of code. I'll open up an issue to create it and replace this implementation with it, but leave it out of this PR.

File: libs/ml/tests/ml/ops/activations/sigmoid.cpp (hunk 11 above)
Comment (baykaner) [reply libs/ml/tests/ml/ops/activations/sigmoid.cpp:61]: python script should show how to generate test values now. In future we should add the python scripts demonstrating how the values are arrived at (whether its numpy, pytorch, tensorflow, or otherwise). Currently I just pushed to scripts/python_ml_lib/ but probably we can tidy up a bit where we keep these.

File: scripts/python_ml_lib/cross_entropy_test.py (hunk 13 above)
Comment (baykaner) [reply scripts/python_ml_lib/cross_entropy_test.py:1]: no; scripting code for validating the values in the unit tests

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp (hunk 2 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:547]: fixed to return Tensor<SizeType> but calling Argmax(array, axis) to return a SizeType now requires the user to be explicit about the templates e.g. SizeArrayType ret = ArgMax<ArrayType, SizeArrayType>(array, axis)

Comment (baykaner) [reply libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:547]: actually on second thought I'm not going to fix this now because it requires adding a new SizeArrayType template to Argmax, but that ends up propagating through most of the ML library annoyingly. The better solution is to rework math and ml library such that instead of just ArrayType template, we give <ArrayType, DataType, SizeType> (or possibly just <DataType, SizeType> in cases where tensor is already included). This is big job so I'm going to bundle this together with removing the shapeless array and do this all in one go. Will add to the shapelessarray issue

Comment (baykaner) [reply libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp:547]: issue #794 regarding this

File: libs/ml/examples/file_loader.hpp (hunk 6 above)
Comment (baykaner) [reply libs/ml/examples/file_loader.hpp]: yup, updated

File: libs/ml/examples/file_loader.hpp [hunk 15]
```diff
@@ -38,18 +38,15 @@ std::vector<std::string> GetAllTextFiles(std::string const &dir_name)
...
     {
-      strtok(ent->d_name, ".");
-      p1 = strtok(NULL, ".");
-      if (p1 != NULL)
+      p1 = ent->d_name;
+      if (p1.find('.'))
       {
-        txt_cmp = std::strcmp(p1, "txt");
-        if (txt_cmp == 0)
+        if (p1.compare("txt"))
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/examples/file_loader.hpp]: That will never evaluate to true, unless the folder contains a file called just "txt"

**Review by pierre-wilmot-fetch-ai — DISMISSED** [review]

File: libs/ml/examples/file_loader.hpp (hunk 15 above)
Comment (baykaner) [reply libs/ml/examples/file_loader.hpp]: 👍

**Review by pierre-wilmot-fetch-ai — APPROVED** [review]
